### PR TITLE
Use explicit destinations in codegen to avoid uselessly jumping around.

### PR DESCRIPTION
**Pull request by sjrd**
Backport of https://github.com/lampepfl/dotty/pull/14890

### CODE REVIEW

**Comment by lrytz** [comment]
So nice, thank you!

File: test/junit/scala/tools/nsc/backend/jvm/opt/InlinerTest.scala [hunk 1]
```diff
@@ -1443,17 +1443,11 @@ class InlinerTest extends BytecodeTesting {
...
-    catch { case e: AssertionError =>
-      try assertSameSummary(getMethod(c, "t"), List( // this is the new behaviour, after restarr'ing
-        ALOAD, "debug", IFEQ /*A*/,
-        "$anonfun$t$2", IRETURN,
-        -1 /*A*/, "$anonfun$t$3", IRETURN))
-      catch { case _: AssertionError => throw e }
-    }
+      "$anonfun$t$2", ISTORE, GOTO /*B*/,
+      -1 /*A*/, "$anonfun$t$3", ISTORE,
+      -1 /*B*/, ILOAD, IRETURN))
```
Comment (sjrd) [inline test/junit/scala/tools/nsc/backend/jvm/opt/InlinerTest.scala]: @lrytz Do you have any idea why the optimizer generates *worse* code after these changes than before? With the changes, with or without optimizer, the bytecode summary generated for `ifelse` is
```
ALOAD, "apply$mcZ$sp", IFEQ /*8*/, ALOAD, "apply", ARETURN, -1 /*8*/, ALOAD, "apply", ARETURN
```
So why does the generated code for `t` regresses to having an `ISTORE`/`GOTO`/`ILOAD`?

Or perhaps a better question: how did the optimizer have the direct `IRETURN`s *before* the changes?

Do you have any pointer in the optimizer to help me figure this out?

Comment (lrytz) [reply test/junit/scala/tools/nsc/backend/jvm/opt/InlinerTest.scala]: It seems the difference in the generated code is that the `store` is pushed into the two branches. before there were jumps to the single store instruction.

`sc12 Test.scala -opt:l:inline '-opt-inline-from:**' -Yopt-trace C.t`


```diff
@@ -65,27 +65,27 @@
     POP
     INVOKESTATIC C.$anonfun$t$2 ()I
     INVOKESTATIC scala/runtime/BoxesRunTime.boxToInteger (I)Ljava/lang/Integer;
+    ASTORE 5
     GOTO L9
    L7
     ALOAD 4
     POP
     INVOKESTATIC C.$anonfun$t$3 ()I
     INVOKESTATIC scala/runtime/BoxesRunTime.boxToInteger (I)Ljava/lang/Integer;
-   L9
     ASTORE 5
-    GOTO L10
-   L11
+    GOTO L9
    L10
+   L9
     ALOAD 5
-    LINENUMBER 4 L10
+    LINENUMBER 4 L9
     INVOKESTATIC scala/runtime/BoxesRunTime.unboxToInt (Ljava/lang/Object;)I
     IRETURN
-   L12
+   L11
```

So the store-load-pair removal no longer triggers

https://github.com/scala/scala/blob/7dbc139b4a113c18e99e8720b985a249fc9b181f/src/compiler/scala/tools/nsc/backend/jvm/opt/CopyProp.scala#L476-L500

Comment (sjrd) [reply test/junit/scala/tools/nsc/backend/jvm/opt/InlinerTest.scala]: Thanks. That sounds plausible indeed! I'll see if I can do something to recover from that.

Comment (sjrd) [reply test/junit/scala/tools/nsc/backend/jvm/opt/InlinerTest.scala]: I pushed a commit in the inliner that addresses this regression.

**Review by lrytz — APPROVED** [review]
LGTM, thank you for all the follow-ups! And sorry about the lie.

File: test/files/neg/inlineMaxSize.scala [hunk 2]
```diff
@@ -5,5 +5,5 @@ class C {
   @inline final def g = f + f + f + f + f + f + f + f + f + f
   @inline final def h = g + g + g + g + g + g + g + g + g + g
   @inline final def i = h + h + h + h + h + h + h + h + h + h
-  @inline final def j = i + i + i
+  @inline final def j = i + i + i + i
```
Comment (sjrd) [inline test/files/neg/inlineMaxSize.scala:8]: I had to adapt that test to make it even bigger for the warning to still appear. :p

**Comment by lrytz** [comment]
Merging so we can see how it goes on the community build.

**Comment by sjrd** [comment]
Would you like me to also backport https://github.com/lampepfl/dotty/pull/15081 ? It won't have as much benefit in scalac as in dotc, but it shouldn't hurt.

**Comment by lrytz** [comment]
Sure, I'd be happy to have these improvements :)

**Comment by SethTisue** [comment]
@sjrd this (and #10025) didn't merge forward onto 2.13 easily; can you submit a forward-port PR?

**Comment by sjrd** [comment]
Probably not this week. But maybe next week? IIUC 2.13.9 is going to wait a bit, isn't it?

**Comment by SethTisue** [comment]
> IIUC 2.13.9 is going to wait a bit, isn't it?

Correct. 2.12.16 is imminent, but 2.13.9 won't ship until after Scala 3.2.0 final is released. So yes, next week or even the week after is soon enough.

**Pull request merged**